src/staker.cairo

```diff
@@ -333,5 +353,67 @@ pub mod Staker {
             let now = get_block_timestamp();
             self.get_average_delegated(delegate, now - period, now)
         }
+
+        fn get_cumulative_seconds_per_total_staked_at(
+            self: @ContractState, timestamp: u64,
+        ) -> u256 {
+            if let Option::Some((log_record, idx)) = self
+                .staking_log
+                .find_in_change_log(timestamp) {
+                let total_staked = if (idx == self.staking_log.len() - 1) {
+                    // if last rescord found
+                    self.total_staked.read()
+                } else {
+                    // otherwise calculate using cumulative_seconds_per_total_staked difference
+                    let next_log_record = self.staking_log.at(idx + 1).read();
+
+                    // substract fixed point values
+                    let divisor = next_log_record.cumulative_seconds_per_total_staked
+                        - log_record.cumulative_seconds_per_total_staked;
+
+                    if divisor.is_zero() {
+                        return 0_u64.into();
+                    }
+
+                    let diff_seconds: u128 = (next_log_record.timestamp - log_record.timestamp)
+                        .into();
+
+                    // Divide u64 by fixed point
+                    let (total_staked_fp_medium, _) = u512_safe_div_rem_by_u256(
+                        u512 { limb0: 0, limb1: 0, limb2: 0, limb3: diff_seconds },
+                        divisor.try_into().unwrap(),
+                    );
+
+                    let total_staked_fp = u256 {
+                        low: total_staked_fp_medium.limb1, high: total_staked_fp_medium.limb2,
+                    };
+
+                    assert(total_staked_fp.high < MAX_FP, 'FP_OVERFLOW');
+
+                    // round value
+                    total_staked_fp.high + if (total_staked_fp.low >= TWO_POW_127) {
+                        1
+                    } else {
+                        0
+                    }
+                };
+
+                let seconds_diff = timestamp - log_record.timestamp;
+
+                let staked_seconds: u256 = if total_staked == 0 {
+                    0_u256
+                } else {
+                    // Divide u64 by u128
+                    u256 { low: 0, high: seconds_diff.into() } / total_staked.into()
+                };
+
+                // Sum fixed posits
+                let result = log_record.cumulative_seconds_per_total_staked + staked_seconds;
+                assert(result.high < MAX_FP, 'FP_OVERFLOW');
```

Comment:
why are you doing this assertion? if it's because you're storing a u256 as a felt252 in the store packing, then it should be asserted where you're actually doing the storage

Reply:
Had same though, but decided to fail as early as possible.

Reply:
Looks like in storage packing it'll fait anyway.